src/phoenix/metrics/__init__.py

```diff
@@ -1,22 +1,62 @@
-from typing import Any, Iterable, Iterator, Mapping, Protocol, runtime_checkable
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Iterable, Iterator, List, Mapping, Optional, Union
 
+import numpy as np
 import pandas as pd
 
+from phoenix.core.model_schema import Column
 
-@runtime_checkable
-class Metric(Protocol):
-    def __call__(self, df: pd.DataFrame) -> Any:
-        ...
 
+@dataclass(frozen=True)
+class Metric(ABC):
     def id(self) -> int:
-        ...
+        """
+        id is a unique identifier for each metric instance. This is used to
+        extract the metric's own value from a collective output containing
+        results from other metrics.
+        """
+        return id(self)
 
-    def input_column_names(self) -> Iterator[str]:
-        ...
+    def initial_value(self) -> Any:
+        return np.nan
```

Comment:
yes, it could be static, but the static signature would preclude the child classes from using their `self` variable, so it would be inflexible in that regard

Reply:
On a second thought, you're indeed correct about the feasibility of it being a static method (or a property), because we're doing mixins for the `initial_value` instead of inheritance. I have updated the it to be property instead, since that's probably the more intuitive of the two. Originally I had a preference for methods over properties because properties can't be picked up by protocols (but we're no longer doing protocols so it's now moot)